File: firmware/display.go
//go:build tinygo && !modemonly

package main

import (
	"image/color"
	"machine"
	"strconv"
	"time"

	"tinygo.org/x/drivers"
	"tinygo.org/x/drivers/st7789"
	"tinygo.org/x/tinyfont"
	"tinygo.org/x/tinyfont/freemono"
	"tinygo.org/x/tinyfont/proggy"
)

const (
	panelW int16 = 135
	panelH int16 = 240
)

// Nokia-ish phosphor LCD palette.
var (
	colBg     = color.RGBA{R: 12, G: 28, B: 14, A: 255}
	colFg     = color.RGBA{R: 170, G: 220, B: 120, A: 255}
	colMuted  = color.RGBA{R: 70, G: 110, B: 60, A: 255}
	colOk     = color.RGBA{R: 190, G: 240, B: 140, A: 255}
	colWarn   = color.RGBA{R: 200, G: 180, B: 60, A: 255}
	colAccent = color.RGBA{R: 150, G: 210, B: 100, A: 255}
	colBarOn  = color.RGBA{R: 180, G: 230, B: 130, A: 255}
	colBarOff = color.RGBA{R: 35, G: 55, B: 35, A: 255}
	colSoft   = color.RGBA{R: 40, G: 70, B: 40, A: 255}
)

type Display struct {
	dev  st7789.Device
	font tinyfont.Fonter
	w, h int16
}

type StatusView struct {
	WiFiOK   bool
	Bars     int
	Operator string
	SMS      int
	Missed   int
	Clock    string
	Alive    int // 0..3 pulse
	HintL    string
	HintR    string
}

func NewDisplay() *Display {
	if err := machine.SPI0.Configure(machine.SPIConfig{
		Frequency: 20_000_000,
		SCK:       pinTFTSCK,
		SDO:       pinTFTMOSI,
		SDI:       machine.NoPin,
		Mode:      0,
	}); err != nil {
		println("spi:", err.Error())
	}
	d := st7789.New(machine.SPI0, pinTFTRST, pinTFTDC, pinTFTCS, pinTFTBL)
	d.Configure(st7789.Config{
		Width:        panelW,
		Height:       panelH,
		Rotation:     drivers.Rotation90,
		RowOffset:    40,
		ColumnOffset: 53,
	})
	disp := &Display{
		dev:  d,
		font: &proggy.TinySZ8pt7b,
	}
	disp.w, disp.h = d.Size()
	disp.dev.FillScreen(colBg)
	return disp
}

func (d *Display) Clear() {
	d.dev.FillScreen(colBg)
}

func (d *Display) Sleep() {
	d.dev.EnableBacklight(false)
}

func (d *Display) Wake() {
	d.dev.EnableBacklight(true)
}

func (d *Display) line(x, y int16, s string, c color.RGBA) {
	tinyfont.WriteLine(&d.dev, d.font, x, y, s, c)
}

func (d *Display) DrawBoot(msg string) {
	d.Clear()
	d.line(8, 20, "NOKIA-ish", colAccent)
	d.line(8, 48, "SMS GATEWAY", colFg)
	if msg == "" {
		msg = "…"
	}
	d.line(8, 80, msg, colMuted)
	d.drawSoftkeys(" ", " ")
}

func (d *Display) DrawStatus(v StatusView) {
	d.Clear()

	drawAntenna(&d.dev, 6, 8, v.Bars)
	wifi := "WIFI-"
	wc := colWarn
	if v.WiFiOK {
		wifi = "WIFI+"
		wc = colOk
	}
	d.line(70, 18, wifi, wc)

	clk := v.Clock
	if clk == "" {
		clk = "--:--"
	}
	_, clkW := tinyfont.LineWidth(d.font, clk)
	d.line(d.w-18-int16(clkW)-6, 16, clk, colFg)
	drawAlive(&d.dev, d.w-18, 10, v.Alive)

	op := v.Operator
	if op == "" {
		op = "—"
	}
	if len(op) > 22 {
		op = op[:22]
	}
	d.line(8, 48, op, colMuted)

	sms := "?"
	if v.SMS >= 0 {
		sms = strconv.Itoa(v.SMS)
	}
	miss := "?"
	if v.Missed >= 0 {
		miss = strconv.Itoa(v.Missed)
	}
	big := &freemono.Bold12pt7b
	d.line(8, 68, "SMS", colMuted)
	d.line(120, 68, "CALLS", colMuted)
	tinyfont.WriteLine(&d.dev, big, 8, 98, sms, colFg)
	tinyfont.WriteLine(&d.dev, big, 120, 98, miss, colFg)

	hl, hr := v.HintL, v.HintR
	if hl == "" {
		hl = "Inbox"
	}
	if hr == "" {
		hr = "Refresh"
	}
	d.drawSoftkeys(hl, hr)
}

func (d *Display) drawSoftkeys(left, right string) {
	y := d.h - 18
	d.dev.FillRectangle(0, y-4, d.w, 22, colSoft)
	d.line(8, y+10, left, colFg)
	// right-align roughly
	rx := d.w - int16(6*len(right)) - 8
	if rx < d.w/2 {
		rx = d.w/2 + 10
	}
	d.line(rx, y+10, right, colFg)
}

func (d *Display) DrawInbox(list []SMS, idx int) {
	d.Clear()
	d.line(8, 16, "INBOX", colAccent)
	if len(list) == 0 {
		d.line(8, 50, "(empty)", colMuted)
		d.drawSoftkeys(" ", "Back")
		return
	}
	if idx < 0 {
		idx = 0
	}
	if idx >= len(list) {
		idx = len(list) - 1
	}
	d.line(90, 16, strconv.Itoa(idx+1)+"/"+strconv.Itoa(len(list)), colMuted)

	start := idx
	if start > 0 && idx == len(list)-1 {
		start = idx - 1
	}
	y := int16(36)
	for i := start; i < len(list) && y < 100; i++ {
		c := colMuted
		prefix := "  "
		if i == idx {
			c = colFg
			prefix = "> "
		}
		from := list[i].From
		if len(from) > 18 {
			from = from[:18]
		}
		d.line(4, y, prefix+from, c)
		y += 12
		preview := list[i].Text
		if len(preview) > 28 {
			preview = preview[:28]
		}
		d.line(14, y, preview, colMuted)
		y += 16
		if i >= start+1 {
			break
		}
	}
	d.drawSoftkeys("Next", "Open")
}

func (d *Display) DrawSMSDetail(s SMS) {
	d.Clear()
	from := s.From
	if len(from) > 26 {
		from = from[:26]
	}
	d.line(8, 16, from, colAccent)

	text := s.Text
	y := int16(36)
	for len(text) > 0 && y < 100 {
		n := 32
		if n > len(text) {
			n = len(text)
		}
		chunk := text[:n]
		if i := indexByte(chunk, '\n'); i >= 0 {
			chunk = text[:i]
			text = text[i+1:]
		} else {
			text = text[n:]
		}
		d.line(8, y, chunk, colFg)
		y += 12
	}
	d.drawSoftkeys(" ", "Back")
}

func drawAntenna(dev *st7789.Device, x, y int16, bars int) {
	// stem
	dev.FillRectangle(x+2, y+2, 3, 14, colBarOn)
	for i := int16(0); i < 4; i++ {
		h := 4 + i*3
		c := colBarOff
		if int(i) < bars {
			c = colBarOn
		}
		dev.FillRectangle(x+8+i*9, y+(16-h), 6, h, c)
	}
}

func drawAlive(dev *st7789.Device, x, y int16, frame int) {
	// heartbeat block — blinks so you see the loop is alive
	c := colMuted
	if frame%2 == 0 {
		c = colOk
	}
	dev.FillRectangle(x, y, 10, 10, c)
	dev.FillRectangle(x+2, y+2, 6, 6, colBg)
	if frame%2 == 0 {
		dev.FillRectangle(x+3, y+3, 4, 4, c)
	}
}

func indexByte(s string, b byte) int {
	for i := 0; i < len(s); i++ {
		if s[i] == b {
			return i
		}
	}
	return -1
}

// formatClockHHMM for status screen.
func formatClockHHMM(t time.Time) string {
	h, m, s := t.Clock()
	// blink colon via odd/even second handled by caller if needed
	_ = s
	return twoDig(h) + ":" + twoDig(m)
}

func twoDig(n int) string {
	if n < 10 {
		return "0" + strconv.Itoa(n)
	}
	return strconv.Itoa(n)
}

File: firmware/src/display.cpp
#include "display.h"
#include "pins.h"
#include "config.h"
#include <SPI.h>
#include <string.h>

// Nokia-ish phosphor (RGB565)
static const uint16_t COL_BG = 0x08E2;      // ~12,28,14
static const uint16_t COL_FG = 0xA6F7;      // ~170,220,120
static const uint16_t COL_MUTED = 0x4567;   // ~70,110,60
static const uint16_t COL_OK = 0xBEF9;      // ~190,240,140
static const uint16_t COL_WARN = 0xC5A7;    // ~200,180,60
static const uint16_t COL_ACCENT = 0x9692;  // ~150,210,100
static const uint16_t COL_BAR_ON = 0xB6F8;
static const uint16_t COL_BAR_OFF = 0x22A6;
static const uint16_t COL_SOFT = 0x2A28;

bool Display::begin() {
  SPI.begin(PIN_TFT_SCK, -1, PIN_TFT_MOSI);
  pinMode(PIN_TFT_BL, OUTPUT);
  digitalWrite(PIN_TFT_BL, HIGH);

  tft_ = new Adafruit_ST7789(&SPI, PIN_TFT_CS, PIN_TFT_DC, PIN_TFT_RST);
  tft_->init(PANEL_W, PANEL_H);
  tft_->setRotation(3);  // landscape 240×135, 180° vs rotation 1
  w_ = tft_->width();
  h_ = tft_->height();
  tft_->fillScreen(COL_BG);
  tft_->setTextWrap(false);
  return true;
}

void Display::clear() { tft_->fillScreen(COL_BG); }

void Display::sleep() { digitalWrite(PIN_TFT_BL, LOW); }

void Display::wake() { digitalWrite(PIN_TFT_BL, HIGH); }

void Display::line(int16_t x, int16_t y, const char *s, uint16_t c) {
  tft_->setCursor(x, y);
  tft_->setTextColor(c, COL_BG);
  tft_->setTextSize(1);
  tft_->print(s);
}

void Display::drawSoftkeys(const char *left, const char *right) {
  // setRotation(3): FB-left = physical right (OK/D1), FB-right = physical left (Next/D0)
  const char *fb_left = right;
  const char *fb_right = left;

  int16_t y = h_ - 18;
  tft_->fillRect(0, y - 4, w_, 22, COL_SOFT);
  line(8, y + 4, fb_left, COL_FG);
  int16_t rx = w_ - (int16_t)(6 * strlen(fb_right)) - 8;
  if (rx < w_ / 2) rx = w_ / 2 + 10;
  line(rx, y + 4, fb_right, COL_FG);
}

void Display::drawAntenna(int16_t x, int16_t y, int bars) {
  tft_->fillRect(x + 2, y + 2, 3, 14, COL_BAR_ON);
  for (int16_t i = 0; i < 4; i++) {
    int16_t hh = 4 + i * 3;
    uint16_t c = (i < bars) ? COL_BAR_ON : COL_BAR_OFF;
    tft_->fillRect(x + 8 + i * 9, y + (16 - hh), 6, hh, c);
  }
}

void Display::drawAlive(int16_t x, int16_t y, int frame) {
  uint16_t c = (frame % 2 == 0) ? COL_OK : COL_MUTED;
  tft_->fillRect(x, y, 10, 10, c);
  tft_->fillRect(x + 2, y + 2, 6, 6, COL_BG);
  if (frame % 2 == 0) tft_->fillRect(x + 3, y + 3, 4, 4, c);
}

void Display::drawBoot(const char *msg) {
  clear();
  line(8, 12, "NOKIA-ish", COL_ACCENT);
  line(8, 36, "SMS GATEWAY", COL_FG);
  line(8, 64, msg && msg[0] ? msg : "…", COL_MUTED);
  drawSoftkeys(" ", " ");
}

void Display::drawStatus(const StatusView &v) {
  clear();
  drawAntenna(6, 4, v.bars);
  line(70, 10, v.wifi_ok ? "WIFI+" : "WIFI-", v.wifi_ok ? COL_OK : COL_WARN);

  const char *clk = v.clock && v.clock[0] ? v.clock : "--:--";
  int16_t clk_x = w_ - 18 - (int16_t)(6 * strlen(clk)) - 6;
  line(clk_x, 8, clk, COL_FG);
  drawAlive(w_ - 18, 6, v.alive);

  char op[24];
  strncpy(op, v.oper && v.oper[0] ? v.oper : "—", sizeof(op) - 1);
  op[sizeof(op) - 1] = 0;
  if (strlen(op) > 22) op[22] = 0;
  line(8, 28, op, COL_MUTED);

  line(8, 44, "QUEUE", COL_MUTED);
  line(120, 44, "CALLS", COL_MUTED);

  char qbuf[8], miss[8];
  snprintf(qbuf, sizeof(qbuf), "%d", v.queue < 0 ? 0 : v.queue);
  if (v.missed >= 0)
    snprintf(miss, sizeof(miss), "%d", v.missed);
  else
    strcpy(miss, "?");

  tft_->setTextSize(2);
  tft_->setTextColor(COL_FG, COL_BG);
  tft_->setCursor(8, 60);
  tft_->print(qbuf);
  tft_->setCursor(120, 60);
  tft_->print(miss);
  tft_->setTextSize(1);

  // IP над softkeys (низ экрана)
  const char *ip = v.ip && v.ip[0] ? v.ip : "—.—.—.—";
  line(8, h_ - 30, ip, COL_MUTED);

  drawSoftkeys(v.hint_l && v.hint_l[0] ? v.hint_l : "Inbox",
               v.hint_r && v.hint_r[0] ? v.hint_r : "Refresh");
}

void Display::drawInbox(const SMS *list, int n, int idx) {
  clear();
  line(8, 8, "INBOX", COL_ACCENT);
  if (n <= 0) {
    line(8, 40, "(empty)", COL_MUTED);
    drawSoftkeys("Back", "Back");
    return;
  }
  if (idx < 0) idx = 0;
  if (idx >= n) idx = n - 1;
  char cnt[16];
  snprintf(cnt, sizeof(cnt), "%d/%d", idx + 1, n);
  line(90, 8, cnt, COL_MUTED);

  int start = idx;
  if (start > 0 && idx == n - 1) start = idx - 1;
  int16_t y = 28;
  for (int i = start; i < n && y < 100; i++) {
    uint16_t c = (i == idx) ? COL_FG : COL_MUTED;
    char row[40];
    snprintf(row, sizeof(row), "%s%.18s", i == idx ? "> " : "  ", list[i].from);
    line(4, y, row, c);
    y += 12;
    char preview[36];
    strncpy(preview, list[i].text, 28);
    preview[28] = 0;
    line(14, y, preview, COL_MUTED);
    y += 16;
    if (i >= start + 1) break;
  }
  drawSoftkeys("Next", "Back");
}

void Display::drawSMSDetail(const SMS &s) {
  clear();
  char from[32];
  strncpy(from, s.from, 26);
  from[26] = 0;
  line(8, 8, from, COL_ACCENT);

  const char *text = s.text;
  int16_t y = 28;
  while (*text && y < 100) {
    char chunk[36];
    size_t n = 0;
    while (text[n] && text[n] != '\n' && n < 32) n++;
    memcpy(chunk, text, n);
    chunk[n] = 0;
    line(8, y, chunk, COL_FG);
    y += 12;
    text += n;
    if (*text == '\n') text++;
  }
  drawSoftkeys("Back", "Back");
}
